hail/src/main/scala/is/hail/expr/types/physical/PSubsetStruct.scala

```diff
@@ -0,0 +1,225 @@
+import is.hail.asm4s.{Code, Settable, SettableBuilder, Value, const}
+import is.hail.expr.ir.{EmitCodeBuilder, EmitMethodBuilder, IEmitCode}
+import is.hail.expr.types.BaseStruct
+import is.hail.expr.types.virtual.TStruct
+import is.hail.utils._
+
+object PSubsetStruct {
+  def apply(ps: PStruct, fieldNames: String*): PSubsetStruct = {
+    val f = fieldNames.map(f => ps.field(f)).toFastIndexedSeq
+    PSubsetStruct(ps, f)
+  }
+}
+
+// PSubsetStruct is a view of some other PStruct
+// Operations it can take:
+// 1) Modify the view by changing which fields from the backing struct it has access to (add/delete, but don't modify backing PStruct)
+// 2) Modify the backing values (setting requiredeness)
+// 3) Modify properties of the fields in the view (rename), which requires the generation of a new backing PStruct
+// Creating entirely new fields seems out of scope
+final case class PSubsetStruct(ps: PStruct, fields: IndexedSeq[PField]) extends PStruct {
+  if (fields == ps.fields) {
+    log.warn("PSubsetStruct used without subsetting input PStruct")
+  }
+
+  val idxMap: IndexedSeq[Int] = fields.zipWithIndex.map { case (f, i) => {
+    // TODO: if we want to allow rename, this can't be true unless we modify the backing PCanonicalStruct
+    // or are ok with index-only match
+    val psField = ps.field(f.name)
+    assert(f == psField)
+    psField.index
+  }
+  }
+
+  val required = ps.required
+
+  override lazy val virtualType = TStruct(fields.map(f => (f.name -> f.typ.virtualType)):_*)
+  override val types: Array[PType] = fields.map(_.typ).toArray
+
+  override def _pretty(sb: StringBuilder, indent: Int, compact: Boolean) {
+    if (compact) {
+      sb.append("PSubsetStruct{")
+      fields.foreachBetween(_.pretty(sb, indent, compact))(sb += ',')
+      sb += '}'
+    } else {
+      if (size == 0)
+        sb.append("Struct { }")
+      else {
+        sb.append("Struct {")
+        sb += '\n'
+        fields.foreachBetween(_.pretty(sb, indent + 4, compact))(sb.append(",\n"))
+        sb += '\n'
+        sb.append(" " * indent)
+        sb += '}'
+      }
+    }
+  }
+
+  override def truncate(newSize: Int): PSubsetStruct =
+    PSubsetStruct(ps, fields.take(newSize))
+
+  override def deleteField(key: String): PStruct = {
+    assert(fieldIdx.contains(key))
+    val index = fieldIdx(key)
+    val newFields = Array.fill[PField](fields.length - 1)(null)
+    for (i <- 0 until index)
+      newFields(i) = fields(i)
+    for (i <- index + 1 until fields.length)
+      newFields(i - 1) = fields(i).copy(index = i - 1)
+    PSubsetStruct(ps, newFields)
+  }
+
+  // TODO: This feels potentially outside the scope of a view of the backing struct.
+  // If we choose to have this, need to decide whether we modify the backing PCanonicalStruct's field names,
+  // or just the view's.
+  // If just the view's, should have a nameMap in addition to idxMap
+  override def rename(m: Map[String, String]): PStruct = ???
```

Comment:
should rename the underlying struct, I think. The purpose of this method is to change field names at compile time, which we can't currently do in IR.